### PR TITLE
VIDEO-3928 Fix LocalVideoSnapshot

### PR DESCRIPTION
**Pull request by PikaJoyce**
JIRA Ticket : [VIDEO-3928](https://issues.corp.twilio.com/browse/VIDEO-3928)

This PR adds a small fix to the Local Video Snapshot example. I have integrated Charlie's suggested fix to utilize the ImageCapture API.

Here is a demo: [Demo](https://f346a9178a82.ngrok.io)

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: examples/localvideosnapshot/src/helpers.js [hunk 1]
```diff
@@ -10,19 +10,30 @@ var Video = require('twilio-video');
...
   });
 }
 
 /**
  * Take snapshot of the local video from the HTMLVideoElement and render it
  * in the HTMLCanvasElement.
  * @param {HTMLVideoElement} video
- * @param {HTMLCanvasElement} canvas
+ * @param {LocalVideoTrack} localVideoTrack
+ * @param {HTMLCanvasElement | HTMLImageElement} snapshot
```
Comment (manjeshbhargav) [inline examples/localvideosnapshot/src/helpers.js:22]:
```suggestion
 * @param {HTMLCanvasElement|HTMLImageElement} snapshot
```

**Review by manjeshbhargav — APPROVED** [review]

File: examples/localvideosnapshot/src/index.js [hunk 2]
```diff
@@ -22,15 +33,20 @@ getSnippet('./helpers.js').then(function(snippet) {
...
   takeSnapshot.onclick = function() {
-    setCanvasSizeToVideo(canvas, video);
-    takeLocalVideoSnapshot(video, canvas);
+    if(window.ImageCapture) {
+      setSnapshotSizeToVideo(img, video);
+      takeLocalVideoSnapshot(video, localVideoTrack, img);
+    } else {
+      setSnapshotSizeToVideo(canvas, video);
+      takeLocalVideoSnapshot(video, localVideoTrack, canvas);
+    }
```
Comment (charliesantos) [inline examples/localvideosnapshot/src/index.js:45]: Maybe simplify this `if else` block?
```js
const el = window.ImageCapture ? img : canvas;
setSnapshotSizeToVideo(el, video);
takeLocalVideoSnapshot(video, localVideoTrack, el);
```

File: examples/localvideosnapshot/src/index.js [hunk 3]
```diff
@@ -22,15 +33,20 @@ getSnippet('./helpers.js').then(function(snippet) {
...
   };
 });
 
 // Resize the canvas to the video size whenever window is resized.
 window.onresize = function() {
-  setCanvasSizeToVideo(canvas, video);
+  window.ImageCapture ? setSnapshotSizeToVideo(img, video) : setSnapshotSizeToVideo(canvas, video);
```
Comment (charliesantos) [inline examples/localvideosnapshot/src/index.js:51]:
```suggestion
setSnapshotSizeToVideo(window.ImageCapture ? img : canvas, video);
```

File: examples/localvideosnapshot/public/index.html [hunk 4]
```diff
@@ -39,7 +39,8 @@ <h4 class="card-title">Local Video</h4>
         <div class="card">
           <div class="card-block">
             <h4 class="card-title">Snapshot</h4>
-            <canvas id="snapshot" width="320" height="240"></canvas>
+            <canvas class="snapshot-canvas hidden" width="320" height="240"></canvas>
```
Comment (charliesantos) [inline examples/localvideosnapshot/public/index.html:42]: We should not hardcode the height and width here or else the canvas image will be downscaled. Instead, set the height and width in JS to match the mediastream's dimensions, then try positioning/resizing the element using css.

File: examples/localvideosnapshot/src/index.js [hunk 5]
```diff
@@ -6,13 +6,24 @@ var helpers = require('./helpers');
...
 var takeSnapshot = document.querySelector('button#takesnapshot');
 var video = document.querySelector('video#videoinputpreview');
 
+// Show image or canvas
+window.onload = function() {
+  if(window.ImageCapture) {
+    img.classList.remove('hidden');
+  } else {
+    canvas.classList.remove('hidden');
+  }
```
Comment (charliesantos) [inline examples/localvideosnapshot/src/index.js:20]: For simplicity
```js
const el = window.ImageCapture ? img : canvas;
el.classList.remove('hidden');
```

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: examples/localvideosnapshot/src/helpers.js [hunk 6]
```diff
@@ -10,19 +10,30 @@ var Video = require('twilio-video');
...
+      const imageCapture = new ImageCapture(localVideoTrack.mediaStreamTrack);
+      imageCapture.takePhoto().then(function(blob) {
+        snapshot.src = URL.createObjectURL(blob);
+        resolve(URL.createObjectURL(blob));
+      });
+    } else {
+      snapshot.getContext('2d').drawImage(video, 0, 0);
+      resolve(snapshot.toDataURL());
+    }
+  });
```
Comment (charliesantos) [inline examples/localvideosnapshot/src/helpers.js:36]: We are calling `takeLocalVideoSnapshot` in index.js but we are not really using the return value. Maybe remove the promise here? Something like below.

```js
function takeLocalVideoSnapshot(video, localVideoTrack, snapshot) {
  if (window.ImageCapture) {
    const imageCapture = new ImageCapture(localVideoTrack.mediaStreamTrack);
    imageCapture.takePhoto().then(function(blob) {
      snapshot.src = URL.createObjectURL(blob);
    });
  } else {
    snapshot.getContext('2d').drawImage(video, 0, 0);
  }
}
```

File: examples/localvideosnapshot/public/index.html (hunk 4 above)
Comment (PikaJoyce) [reply examples/localvideosnapshot/public/index.html]: This introduced some of the styling on safari to look funky. Going to fiddle with the CSS to make it look nice! Thanks Charlie!!!

**Review by charliesantos — APPROVED** [review]

**Pull request merged**